Reworked Attributes tab navigation

LemonUnit123 opened the pull request:
- Don't hide it behind research (removed Requirements for attributesTabButton in js\game-config\requirements.js)
- make attributes instantly clickable to show attributes tab (changed in js\main.js by removing attributesTabButton from tabButtons and adding attributesDisplay to tabButtons)
- when clicking attributes tab again, close attributes tab (added logic in js\main.js in function setTab(selectedTab){..})
- remove attributes tab from the side bar (in index.html: added tabButton to class and removed attributesTabButtonRequirements + attributesTabButton)

Did some additional changes in css\global\_main.css for visual appearance and js\main.js in function setTab(selectedTab){..} for horizontal Animation (AttributesTab) and function updateConnector(){..} for the vertical connection line between active element and content

LemonUnit123 committed:
Reworked Attributes tab navigation

- Don't hide it behind research (removed Requirements for attributesTabButton in js\game-config\requirements.js)
- make attributes instantly clickable to show attributes tab (changed in js\main.js by removing attributesTabButton from tabButtons and adding attributesDisplay to tabButtons)
- when clicking attributes tab again, close attributes tab (added logic in js\main.js in function setTab(selectedTab){..})
- remove attributes tab from the side bar (in index.html: added tabButton to class and removed attributesTabButtonRequirements + attributesTabButton)

Did some additional changes in css\global\_main.css for visual appearance and js\main.js in function setTab(selectedTab){..} for horizontal Animation (AttributesTab) and function updateConnector(){..} for the vertical connection line between active element and content

r-zander commented on index.html:
```diff
@@ -110,7 +110,7 @@ <h2 id="nameDisplay" class="shortcut my-0 px-3 pb-1 text-nowrap overflow-x-hidde
 
         <div id="attributesDisplay"
-             class="shortcut panel px-3 py-1 d-flex justify-content-start gap-4 overflow-x-auto"
+             class="shortcut tabButton panel px-3 py-1 d-flex justify-content-start gap-4 overflow-x-auto"
```
Mhm, .shortcut und .tabButton haben vorher schon fast dasselbe gemacht - jetzt hast du #attributesDisplay beide Klassen gegeben, meinst du wir könnten einfach eine davon los werden?

r-zander replied:
Nachdem ich jetzt den Rest vom Code (vor allem die Connector Logic & Tab Logic) nochmal gelesen habe... ja, ich würde .shortcut einfach droppen und die nötigen stylings für `#attributesDisplay.shortcut` in `#attributesDisplay.tabButton` bewegen.

r-zander commented on js/main.js:
```diff
@@ -80,20 +89,47 @@ function setTab(selectedTab) {
+    if(isVerticalBarAnimation) {
+        content.style.borderTop = 'none';
+        content.style.borderLeft = '3px solid var(--active-color)';
+        animBar.className = 'bar-animation vertical';
```
```suggestion
        animBar.classList.add('vertical');
        animBar.classList.remove('horizontal');
```
ist netter als das gesamte `class` attribute neu zu schreiben.

Kannst du entsprechend auch bei horizontal anwenden.

r-zander replied:
Das'  jetzt alles etwas wild.

An sich mag ich die Idee, ein eigenes Element für diese kleine Animation zu haben. Andererseits sind Elemente, die rein dekorativ sind eigentlich zu vermeiden, und `#content` hatte auch bereits quasi alles dafür aufgesetzt. 
Mein Vorschlag: nimm das .bar-animation Element weg, und gestalte `#content` mit `.vertical-bar` bzw. `.horizontal-bar`. 
Dann kannst du hier wahrscheinlich auch die Style-Manipulationen vermeiden.

r-zander commented on js/main.js:
```diff
@@ -124,46 +160,57 @@ function updateConnector() {
 
-    const margin = 16;
-    const connectorHeight = 3;
+    if (!activeTabButton || !connectorElement || !contentElement) return;
```
Etwas, was ich jedem in JavaScript beibringe und sicherlich auch streitbar ist, but hear me out.

1. Vertraue deinem (unserem :D ) Code. Dinge sollten nicht random null/undefined/falsy sein. Wenn du den Code im Griff hast (durch Debuggen und Docu lesen), **weißt** du welche Werte hier zu erwarten sind.
2. Keine allgemeinen truthy oder falsy checks. Wenn notwendig (hier ist es das imho nicht - diese Elemente sollten vorhanden sein, sonst ist irgendwas komisch), checke mit is[Whatever] ob die Variable den erwarteten Wert hat. Bzw wenn die Variable, bei so Selektoren, exakt entweder `null` oder ein HTMLElement ist, dann passt auch ein `variable !== null` um die Existenz zu checken.

Warum das ganze? 
- Es verhindert, dass man alles überall gegen alles checkt - und trotzdem eigentlich sicher sein kann, dass die Variable den erwarteten Typ hat
- Es gibt auch den anderen Entwicklern vertrauen in den Code. Wenn man einmal mit solchen broad truthy/falsy checks anfängt, breitet sich das überall aus, weil niemand mehr schnell sehen kann, von welchem Typ eigentlich welche Variable ist.

r-zander commented on js/main.js:
```diff
@@ -124,46 +160,57 @@ function updateConnector() {
+            connectorElement.style.top = `${Math.round(yStart)}px`;
+            connectorElement.style.height = `${Math.round(yEnd - yStart)}px`;
+            connectorElement.style.width = `${connectorThickness}px`;
+        }
```
Props das so umzuschreiben 👏 
Scheint auch zu funktionieren 🔥

r-zander commented on js/main.js:
```diff
@@ -80,20 +89,47 @@ function setTab(selectedTab) {
+        );
+      
+    }
+    else {
```
```suggestion
    } else {
```

Bitte auf Formatierung achten.

r-zander commented on js/main.js:
```diff
@@ -80,20 +89,47 @@ function setTab(selectedTab) {
+    const content = Dom.get().byId('content');
+    const animBar = content.querySelector('.bar-animation');
+
+    if(isVerticalBarAnimation) {
```
```suggestion
    if (isVerticalBarAnimation) {
```

r-zander reviewed (changes requested):
Ja crazy, sieht richtig gut aus!
Paar Kleinigkeiten zum Code-Style/Consistency aber nichts dramatisches.

LemonUnit123 committed:
Changes for PullRequest (Reworked Attributes tab navigation)

- removed shortcut from attributesDisplay in index.html
- removed bar-animation element from index.html
- removed .bar-animation from _main.css and added .horizontal and .vertical to #content
- removed animBar and using content element instead in main.js for animation
- removed unnecessary null check

r-zander commented on index.html:
```diff
@@ -110,7 +110,7 @@ <h2 id="nameDisplay" class="shortcut my-0 px-3 pb-1 text-nowrap overflow-x-hidde
 
         <div id="attributesDisplay"
-             class="shortcut tabButton panel px-3 py-1 d-flex justify-content-start gap-4 overflow-x-auto"
+             class="tabButton panel px-3 py-1 d-flex justify-content-start gap-4 overflow-x-auto"
```
👍 

Zwei Dinge fehlen leider noch:
- .shortcut wird auch im Code gesetzt
- .shortcut hatte noch CSS properties drin, genau für solche DIVs die als Button erscheinen sollten. Die würde ich einfach auf `#attributesDisplay.tabButton` auch anwenden

LemonUnit123 replied:
.shortcut wird jetzt für attributesDisplay nicht mehr im code gesetzt und ich habe im _navigation.css .shortcut zu #attributesDisplay.tabButton geändert

r-zander commented on css/global/_main.css:
```diff
@@ -137,25 +137,12 @@ a#gameTitle {
+#content.vertical {
     background-repeat: repeat-y;
     background-size: 0 100%;
 }
```
```suggestion
    border-left: 3px solid var(--active-color);

    &.horizontal {
        background-repeat: repeat-x;
        background-size: 100% 0;
    }
    
    #content.vertical {
        background-repeat: repeat-y;
        background-size: 0 100%;
    }
}
```

Du kannst nested selectors verwenden, siehe https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector

LemonUnit123 replied:
Das ist gut zu wissen, danke. Wird so umgesetzt.

r-zander commented on js/main.js:
```diff
@@ -90,13 +90,11 @@ function setTab(selectedTab) {
 
-    if(isVerticalBarAnimation) {
-        content.style.borderTop = 'none';
-        content.style.borderLeft = '3px solid var(--active-color)';
```
Dieser Style fehlt noch im CSS - so ist aktuell die Highlight-Border auf der falschen Seite

<img width="1537" height="664" alt="Image" src="https://github.com/user-attachments/assets/3cc7478c-68c8-462d-8c7e-836702d35c43" />

LemonUnit123 committed:
Changes (2) for PullRequest (Reworked Attributes tab navigation)

- removed shortcut assignment to attributesDisplay from js\main.js
- using nested selectors for #content.horizontal and #content.vertical in css\global\_main.css
- renamed .shortcut to #attributesDisplay.tabButton in css\elements\_navigation.css

r-zander committed:
Fixed CSS issue.

r-zander committed:
Merge branch 'master' into dev-leon

# Conflicts:
#	js/game-config/requirements.js

r-zander reviewed (approved):
Very cool, thank you!
